lib/Dialect/SPIRV/SPIRVOps.cpp

```diff
@@ -449,6 +449,34 @@ static void printLogicalOp(Operation *logicalOp, OpAsmPrinter &printer) {
   printer << " : " << logicalOp->getOperand(0)->getType();
 }
 
+static ParseResult parseShiftOp(OpAsmParser &parser, OperationState &state) {
+  SmallVector<OpAsmParser::OperandType, 2> operandInfo;
+  Type baseType;
+  Type shiftType;
+  auto loc = parser.getCurrentLocation();
+
+  if (parser.parseOperandList(operandInfo, 2) || parser.parseColon() ||
+      parser.parseType(baseType) || parser.parseComma() ||
+      parser.parseType(shiftType) ||
+      parser.resolveOperands(operandInfo, {baseType, shiftType}, loc,
+                             state.operands)) {
+    return failure();
+  }
+  state.addTypes(baseType);
+  return success();
+}
+
+static void printShiftOp(Operation *op, OpAsmPrinter &printer) {
+  Value *base = op->getOperand(0);
+  Value *shift = op->getOperand(1);
+  printer << op->getName() << ' ' << *base << ", " << *shift << " : "
+          << base->getType() << ", " << shift->getType();
+}
+
+static LogicalResult verifyShiftOp(Operation *op) {
+  return success(op->getOperand(0)->getType() == op->getResult(0)->getType());
```

Comment:
Print an error message on failure and also add a test for this? :)

Reply:
@antiagainst oh, sorry, I forgot about an error message;
by the way, it seems like an error could be generated only when manually creating this op with builder, not from parser, because while parsing it reuses `baseType` https://github.com/tensorflow/mlir/pull/215/files#diff-1a7b8ff4af6a0a814aa170d35a45ba8bR465
please fix me if I'm wrong.
Thanks.

Reply:
Yes you are right. Please add a test with the generic assembly form. Thanks! :)

Reply:
@antiagainst thanks for mention generic assembly form! I didn't know about it.